File: src/ClientModel/ProxyMethod.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text.RegularExpressions;
using AutoRest.Core.Model;
using AutoRest.Core.Utilities;
using AutoRestMethod = AutoRest.Core.Model.Method;

namespace AutoRest.Java.Model
{
    /// <summary>
    /// A method within a Proxy.
    /// </summary>
    public class ProxyMethod
    {
        /// <summary>
        /// Create a new RestAPIMethod with the provided properties.
        /// </summary>
        /// <param name="requestContentType">The Content-Type of the request.</param>
        /// <param name="returnType">The type of value that is returned from this method.</param>
        /// <param name="isPagingNextOperation">Whether or not this method is a request to get the next page of a sequence of pages.</param>
        /// <param name="httpMethod">The HTTP method that will be used for this method.</param>
        /// <param name="urlPath">The path of this method's request URL.</param>
        /// <param name="responseExpectedStatusCodes">The status codes that are expected in the response.</param>
        /// <param name="returnValueWireType">The return value's type as it is received from the network (across the wire).</param>
        /// <param name="unexpectedResponseExceptionType">The exception type to throw if this method receives and unexpected response status code.</param>
        /// <param name="name">The name of this REST API method.</param>
        /// <param name="asyncReturnType">The return type of this method with its asynchronous container.</param>
        /// <param name="parameters">The parameters that are provided to this method.</param>
        /// <param name="description">The description of this method.</param>
        /// <param name="returnValueClientType">The return value's type as it is returned from the client.</param>
        /// <param name="autoRestMethod">The AutoRestMethod that this RestAPIMethod was created from.</param>
        /// <param name="isResumable">Whether or not this method is resumable.</param>
        public ProxyMethod(string requestContentType, IType returnType, bool isPagingNextOperation, HttpMethod httpMethod, string urlPath, IEnumerable<HttpStatusCode> responseExpectedStatusCodes, ClassType unexpectedResponseExceptionType, string name, IEnumerable<ProxyMethodParameter> parameters, string description, IType returnValueWireType, MethodJv autoRestMethod,
            bool isResumable)
        {
            RequestContentType = requestContentType;
            ReturnType = returnType;
            IsPagingNextOperation = isPagingNextOperation;
            HttpMethod = httpMethod;
            UrlPath = urlPath;
            ResponseExpectedStatusCodes = responseExpectedStatusCodes;
            UnexpectedResponseExceptionType = unexpectedResponseExceptionType;
            Name = name;
            Parameters = parameters;
            Description = description;
            ReturnValueWireType = returnValueWireType;
            AutoRestMethod = autoRestMethod;
            IsResumable = isResumable;
        }

        /// <summary>
        /// Get the Content-Type of the request.
        /// </summary>
        public string RequestContentType { get; }

        /// <summary>
        /// The value that is returned from this method.
        /// </summary>
        public IType ReturnType { get; }

        /// <summary>
        /// Get whether or not this method is a request to get the next page of a sequence of pages.
        /// </summary>
        public bool IsPagingNextOperation { get; }

        /// <summary>
        /// Get the HTTP method that will be used for this method.
        /// </summary>
        public HttpMethod HttpMethod { get; }

        /// <summary>
        /// Get the path of this method's request URL.
        /// </summary>
        public string UrlPath { get; }

        /// <summary>
        /// Get the status codes that are expected in the response.
        /// </summary>
        public IEnumerable<HttpStatusCode> ResponseExpectedStatusCodes { get; }

        /// <summary>
        /// Get the exception type to throw if this method receives and unexpected response status code.
        /// </summary>
        public ClassType UnexpectedResponseExceptionType { get; }

        /// <summary>
        /// Get the name of this Rest API method.
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// Get the parameters that are provided to this method.
        /// </summary>
        public IEnumerable<ProxyMethodParameter> Parameters { get; }

        /// <summary>
        /// Get the description of this method.
        /// </summary>
        public string Description { get; }

        /// <summary>
        /// The value of the ReturnValueWireType annotation for this method.
        /// </summary>
        public IType ReturnValueWireType { get; }

        /// <summary>
        /// The AutoRestMethod that this RestAPIMethod was created from.
        /// </summary>
        public MethodJv AutoRestMethod { get; }

        /// <summary>
        /// Get whether or not this method resumes polling of an LRO.
        /// </summary>
        public bool IsResumable { get; }

        public string PagingAsyncSinglePageMethodName => Name + "SinglePageAsync";

        public string SimpleAsyncMethodName => Name + "Async";

        public string SimpleAsyncRestResponseMethodName => Name + "WithRestResponseAsync";

        private MethodType? _methodType;
        public MethodType MethodType
        {
            get
            {
                if (_methodType != null)
                {
                    return _methodType.Value;
                }
                _methodType = MethodType.Other;
                if (!string.IsNullOrEmpty(AutoRestMethod.MethodGroup?.Name?.ToString()))
                {
                    string autoRestMethodUrl = new Regex("/+$").Replace(new Regex("^/+").Replace(UrlPath, ""), "");
                    string[] autoRestMethodUrlSplits = autoRestMethodUrl.Split('/');
                    switch (AutoRestMethod.HttpMethod)
                    {
                        case HttpMethod.Get:
                            if ((autoRestMethodUrlSplits.Length == 5 || autoRestMethodUrlSplits.Length == 7)
                                && autoRestMethodUrlSplits[0].EqualsIgnoreCase("subscriptions")
                                && AutoRestMethod.ReturnType.Body.MethodHasSequenceType())
                            {
                                if (autoRestMethodUrlSplits.Length == 5)
                                {
                                    if (autoRestMethodUrlSplits[2].EqualsIgnoreCase("providers"))
                                    {
                                        _methodType = MethodType.ListBySubscription;
                                    }
                                    else
                                    {
                                        _methodType = MethodType.ListByResourceGroup;
                                    }
                                }
                                else if (autoRestMethodUrlSplits[2].EqualsIgnoreCase("resourceGroups"))
                                {
                                    _methodType = MethodType.ListByResourceGroup;
                                }
                            }
                            else if (autoRestMethodUrlSplits.IsTopLevelResourceUrl())
                            {
                                _methodType = MethodType.Get;
                            }
                            break;

                        case HttpMethod.Delete:
                            if (autoRestMethodUrlSplits.IsTopLevelResourceUrl())
                            {
                                _methodType = MethodType.Delete;
                            }
                            break;
                    }
                }
                return _methodType.Value;
            }
        }

        /// <summary>
        /// Add this property's imports to the provided ISet of imports.
        /// </summary>
        /// <param name="imports">The set of imports to add to.</param>
        /// <param name="includeImplementationImports">Whether or not to include imports that are only necessary for method implementations.</param>
        public void AddImportsTo(ISet<string> imports, bool includeImplementationImports, JavaSettings settings)
        {
            if (UnexpectedResponseExceptionType != null)
            {
                if (includeImplementationImports)
                {
                    imports.Add("com.azure.common.annotations.UnexpectedResponseExceptionType");
                }

                UnexpectedResponseExceptionType.AddImportsTo(imports, includeImplementationImports);
            }

            if (includeImplementationImports)
            {
                if (IsResumable)
                {
                    imports.Add("com.azure.common.annotations.ResumeOperation");
                }
                imports.Add($"com.azure.common.annotations.{HttpMethod.ToString().ToUpperInvariant()}");

                imports.Add("com.azure.common.annotations.ExpectedResponses");

                if (ReturnValueWireType != null)
                {
                    imports.Add("com.azure.common.annotations.ReturnValueWireType");
                    ReturnValueWireType.AddImportsTo(imports, includeImplementationImports);
                }

                ReturnType.AddImportsTo(imports, includeImplementationImports);

                foreach (ProxyMethodParameter parameter in Parameters)
                {
                    parameter.AddImportsTo(imports, includeImplementationImports, settings);
                }
            }
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/models/clientmodel/ProxyMethod.java
//====================================================================================================
//The Free Edition of C# to Java Converter limits conversion output to 100 lines per file.

//To subscribe to the Premium Edition, visit our website:
//https://www.tangiblesoftwaresolutions.com/order/order-csharp-to-java.html
//====================================================================================================

package com.azure.autorest.models.clientmodel;

import com.azure.autorest.extension.base.plugin.JavaSettings;
import com.azure.autorest.models.codemodel.HttpMethod;
import com.azure.autorest.models.codemodel.HttpOperation;
import jdk.nashorn.internal.runtime.regexp.joni.Regex;

import java.util.List;
import java.util.Set;

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
//C# TO JAVA CONVERTER NOTE: There is no Java equivalent to C# namespace aliases:
//using AutoRestMethod = AutoRest.Core.Model.Method;

/** 
 A method within a Proxy.
*/
public class ProxyMethod
{
	/** 
	 Create a new RestAPIMethod with the provided properties.
	 
	 @param requestContentType The Content-Type of the request.
	 @param returnType The type of value that is returned from this method.
	 @param isPagingNextOperation Whether or not this method is a request to get the next page of a sequence of pages.
	 @param httpMethod The HTTP method that will be used for this method.
	 @param urlPath The path of this method's request URL.
	 @param responseExpectedStatusCodes The status codes that are expected in the response.
	 @param returnValueWireType The return value's type as it is received from the network (across the wire).
	 @param unexpectedResponseExceptionType The exception type to throw if this method receives and unexpected response status code.
	 @param name The name of this REST API method.
	 @param asyncReturnType The return type of this method with its asynchronous container.
	 @param parameters The parameters that are provided to this method.
	 @param description The description of this method.
	 @param returnValueClientType The return value's type as it is returned from the client.
	 @param autoRestMethod The AutoRestMethod that this RestAPIMethod was created from.
	 @param isResumable Whether or not this method is resumable.
	*/
	public ProxyMethod(String requestContentType, IType returnType, boolean isPagingNextOperation, HttpMethod httpMethod, String urlPath, List<HttpStatusCode> responseExpectedStatusCodes, ClassType unexpectedResponseExceptionType, String name, List<ProxyMethodParameter> parameters, String description, IType returnValueWireType, HttpOperation autoRestMethod, boolean isResumable)
	{
		RequestContentType = requestContentType;
		ReturnType = returnType;
		IsPagingNextOperation = isPagingNextOperation;
		HttpMethod = httpMethod;
		UrlPath = urlPath;
		ResponseExpectedStatusCodes = responseExpectedStatusCodes;
		UnexpectedResponseExceptionType = unexpectedResponseExceptionType;
		Name = name;
		Parameters = parameters;
		Description = description;
		ReturnValueWireType = returnValueWireType;
		AutoRestMethod = autoRestMethod;
		IsResumable = isResumable;
	}

	/** 
	 Get the Content-Type of the request.
	*/
	private String RequestContentType;
	public final String getRequestContentType()
	{
		return RequestContentType;
	}

	/** 
	 The value that is returned from this method.
	*/
	private IType ReturnType;
	public final IType getReturnType()
	{
		return ReturnType;
	}

	/** 
	 Get whether or not this method is a request to get the next page of a sequence of pages.
	*/
	private boolean IsPagingNextOperation;
	public final boolean getIsPagingNextOperation()
	{
		return IsPagingNextOperation;
	}

	/** 
	 Get the HTTP method that will be used for this method.
	*/
	private HttpMethod HttpMethod;
	public final HttpMethod getHttpMethod()
	{
		return HttpMethod;
	}

	/** 
	 Get the path of this method's request URL.
	*/
	private String UrlPath;
	public final String getUrlPath()
	{
		return UrlPath;
	}

	/** 
	 Get the status codes that are expected in the response.
	*/
	private List<HttpStatusCode> ResponseExpectedStatusCodes;
	public final List<HttpStatusCode> getResponseExpectedStatusCodes()
	{
		return ResponseExpectedStatusCodes;
	}

	/** 
	 Get the exception type to throw if this method receives and unexpected response status code.
	*/
	private ClassType UnexpectedResponseExceptionType;
	public final ClassType getUnexpectedResponseExceptionType()
	{
		return UnexpectedResponseExceptionType;
	}

	/** 
	 Get the name of this Rest API method.
	*/
	private String Name;
	public final String getName()
	{
		return Name;
	}

	/** 
	 Get the parameters that are provided to this method.
	*/
	private List<ProxyMethodParameter> Parameters;
	public final List<ProxyMethodParameter> getParameters()
	{
		return Parameters;
	}

	/** 
	 Get the description of this method.
	*/
	private String Description;
	public final String getDescription()
	{
		return Description;
	}

	/** 
	 The value of the ReturnValueWireType annotation for this method.
	*/
	private IType ReturnValueWireType;
	public final IType getReturnValueWireType()
	{
		return ReturnValueWireType;
	}

	/** 
	 The AutoRestMethod that this RestAPIMethod was created from.
	*/
	private HttpOperation AutoRestMethod;
	public final HttpOperation getAutoRestMethod()
	{
		return AutoRestMethod;
	}

	/** 
	 Get whether or not this method resumes polling of an LRO.
	*/
	private boolean IsResumable;
	public final boolean getIsResumable()
	{
		return IsResumable;
	}

	public final String getPagingAsyncSinglePageMethodName()
	{
		return getName() + "SinglePageAsync";
	}

	public final String getSimpleAsyncMethodName()
	{
		return getName() + "Async";
	}

	public final String getSimpleAsyncRestResponseMethodName()
	{
		return getName() + "WithRestResponseAsync";
	}

//	private MethodType _methodType = null;
//	public final MethodType getMethodType()
//	{
//		if (_methodType != null)
//		{
//			return _methodType.getValue();
//		}
//		_methodType = getMethodType().Other;
//		if (!tangible.StringHelper.isNullOrEmpty(getAutoRestMethod().MethodGroup == null ? null : (getAutoRestMethod().MethodGroup.Name == null ? null : getAutoRestMethod().MethodGroup.Name.toString())))
//		{
//			String autoRestMethodUrl = (new Regex("/+$")).Replace((new Regex("^/+")).Replace(getUrlPath(), ""), "");
//			String[] autoRestMethodUrlSplits = autoRestMethodUrl.split("[/]", -1);
//			switch (getAutoRestMethod().HttpMethod)
//			{
//				case getHttpMethod().Get:
//					if ((autoRestMethodUrlSplits.length == 5 || autoRestMethodUrlSplits.length == 7) && autoRestMethodUrlSplits[0].EqualsIgnoreCase("subscriptions") && getAutoRestMethod().ReturnType.Body.MethodHasSequenceType())
//					{
//						if (autoRestMethodUrlSplits.length == 5)
//						{
//							if (autoRestMethodUrlSplits[2].EqualsIgnoreCase("providers"))
//							{
//								_methodType = getMethodType().ListBySubscription;
//							}
//							else
//							{
//								_methodType = getMethodType().ListByResourceGroup;
//							}
//						}
//						else if (autoRestMethodUrlSplits[2].EqualsIgnoreCase("resourceGroups"))
//						{
//							_methodType = getMethodType().ListByResourceGroup;
//						}
//					}
//					else if (autoRestMethodUrlSplits.IsTopLevelResourceUrl())
//					{
//						_methodType = getMethodType().Get;
//					}
//					break;
//
//				case getHttpMethod().Delete:
//					if (autoRestMethodUrlSplits.IsTopLevelResourceUrl())
//					{
//						_methodType = getMethodType().Delete;
//					}
//					break;
//			}
//		}
//		return _methodType.getValue();
//	}

	/** 
	 Add this property's imports to the provided ISet of imports.
	 
	 @param imports The set of imports to add to.
	 @param includeImplementationImports Whether or not to include imports that are only necessary for method implementations.
	*/
	public final void AddImportsTo(Set<String> imports, boolean includeImplementationImports, JavaSettings settings)
	{

		if (includeImplementationImports)
		{
			if (getUnexpectedResponseExceptionType() != null)
			{
				imports.add("com.azure.core.implementation.annotation.UnexpectedResponseExceptionType");
				getUnexpectedResponseExceptionType().AddImportsTo(imports, includeImplementationImports);
			}
			if (getIsResumable())
			{
				imports.add("com.azure.core.implementation.annotation.ResumeOperation");
			}
			imports.add(String.format("com.azure.core.implementation.annotation.%1$s", getHttpMethod().toString().ToPascalCase()));

			imports.add("com.azure.core.implementation.annotation.ExpectedResponses");

			if (getReturnValueWireType() != null)
			{

//====================================================================================================
//End of the allowed output for the Free Edition of C# to Java Converter.

//To subscribe to the Premium Edition, visit our website:
//https://www.tangiblesoftwaresolutions.com/order/order-csharp-to-java.html
//====================================================================================================
